src/anomalib/models/components/base/buffer_list.py

```diff
@@ -0,0 +1,108 @@
+"""Buffer List Mixin."""
+
+# Copyright (C) 2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+
+import torch
+from torch import nn
+
+
+class BufferListMixin(nn.Module):
+    """Buffer List Mixin.
+
+    This mixin is used to allow registering a list of tensors as buffers in a pytorch module.
+    """
+
+    def register_bufferlist(self, name: str, values: list[torch.Tensor], persistent: bool = True, **kwargs) -> None:
+        """Register a list of tensors as buffers in a pytorch module.
+
+        Each tensor is registered as a buffer with the name `_name_i` where `i` is the index of the tensor in the list.
+        To update and retrieve the list of tensors, we dynamically assign a descriptor attribute to the class.
+
+        Args:
+            name (str): Name of the bufferlist.
+            values (list[torch.Tensor]): List of tensors to register as buffers.
+            persistent (bool, optional): Whether the buffers should be saved as part of the module state_dict.
+                Defaults to True.
+            **kwargs: Additional keyword arguments to pass to `torch.nn.Module.register_buffer`.
+
+        Example:
+            >>> class MyModule(BufferListMixin, nn.Module):
+            ...     def __init__(self):
+            ...         super().__init__()
+            ...         tensor_list = [torch.ones(3) * i for i in range(3)]
+            ...         self.register_bufferlist("my_bufferlist", tensor_list)
+            >>> module = MyModule()
+            >>> # The bufferlist can be accessed as a regular attribute
+            >>> module.my_bufferlist
+            [
+                tensor([0., 0., 0.]),
+                tensor([1., 1., 1.]),
+                tensor([2., 2., 2.])
+            ]
+            >>> # We can update the bufferlist at any time
+            >>> new_tensor_list = [torch.ones(3) * i + 10 for i in range(3)]
+            >>> module.register_bufferlist("my_bufferlist", new_tensor_list)
+            >>> module.my_bufferlist
+            [
+                tensor([10., 10., 10.]),
+                tensor([11., 11., 11.]),
+                tensor([12., 12., 12.])
+            ]
+            >>> # Move to GPU. Since the tensors are registered as buffers, device placement is handled automatically
+            >>> module.cuda()
+            >>> module.my_bufferlist
+            [
+                tensor([10., 10., 10.], device='cuda:0'),
+                tensor([11., 11., 11.], device='cuda:0'),
+                tensor([12., 12., 12.], device='cuda:0')
+            ]
+        """
+        for i, value in enumerate(values):
+            self.register_buffer(f"_{name}_{i}", value, persistent=persistent, **kwargs)
+
+        setattr(BufferListMixin, name, BufferListDescriptor(name, len(values)))
```

Comment:
What's the advantage of setting this to `BufferListMixin` rather than `self`?

Reply:
It's because descriptors need to be assigned to the class, not the instance, in order to work properly.
see https://stackoverflow.com/questions/12599972/descriptors-as-instance-attributes-in-python
and https://docs.python.org/3/howto/descriptor.html#closing-thoughts